### DISCUSSION TITLE
Error in testing a function

### DISCUSSION THREAD

**Discussion by achyuta22** [Cyfrin/foundry-full-course-cu#2495] [Q&A]
``` Testing script```
// SPDX-License-Identifier:MIT
pragma solidity ^0.8.18;

import {Test} from "../../lib/forge-std/src/Test.sol";
import {DeployRaffle} from "../../script/DeployRaffle.s.sol";
import {Raffle} from "../../src/Raffle.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";

contract RaffleTest is Test {
    Raffle public raffle;
    HelperConfig public helperConfig;
    address public PLAYER = makeAddr("player");
    uint256 constant STARTING_PLAYER_BALANCE = 10 ether;
    uint entranceFee;
    uint interval;
    address vrfCoordinator;
    bytes32 gasLane;
    uint256 subscriptionId;
    uint32 callbackGasLimit;

    function setUp() public {
        DeployRaffle deployer = new DeployRaffle();
        (raffle, helperConfig) = deployer.deployRaffle();
        HelperConfig.NetworkConfig memory config = helperConfig
            .getNetworkByChainId(block.chainid);
        entranceFee = config.entranceFee;
        interval = config.interval;
        vrfCoordinator = config.vrfCoordinator;
        gasLane = config.gasLane;
        subscriptionId = config.subscriptionId;
        callbackGasLimit = config.callbackGasLimit;
        vm.deal(payable(PLAYER), STARTING_PLAYER_BALANCE);

    }

    function testRaffleInitialisesInOpenState() public {
        assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN);
    }

    function testRaffleRevertWhenNotPaidEnough() public {
        vm.deal(payable(PLAYER), STARTING_PLAYER_BALANCE);

        // vm.expectRevert(Raffle.Raffle__SendMoreToEnterRaffle.selector);

        //to send from specific address
        vm.prank(PLAYER);
        // as we are expecting a revert
        //  payable(address(raffle)).transfer(0.02 ether);
       raffle.enterRaffle{value:10 ether}();
        
    }

    function testRaffleRecordsWhenPlayersEntersRaffle() public {

        vm.prank(PLAYER);
        raffle.enterRaffle{value:0.02 ether}();
        assert(raffle.getRecentEntered()==PLAYER);
     
    }

    function testEnteringRaffleEmitEvent() public {
        vm.prank(PLAYER);
        vm.expectEmit(true,false,false,false,address(raffle));
        emit Raffle.RaffleEntered(PLAYER);
        raffle.enterRaffle{value:0.02 ether}();

    }

    function testDontAllowWhileCalculating() public{
        vm.warp(block.timestamp+interval+1);
         vm.roll(block.number+1);
        raffle.performUpkeep("");
        assert(raffle.getRaffleState()==Raffle.RaffleState.CALCULATING);
    }
}

``` RAFFLE.SOL```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

//CEI METHOD
// CHECKS, EFFECTS , INTERACTIONS

import {VRFConsumerBaseV2Plus} from "@chainlink/contracts/src/v0.8/vrf/dev/VRFConsumerBaseV2Plus.sol";
import {VRFV2PlusClient} from "@chainlink/contracts/src/v0.8/vrf/dev/libraries/VRFV2PlusClient.sol";
import {AutomationCompatibleInterface} from "@chainlink/contracts/src/v0.8/automation/AutomationCompatible.sol";

contract Raffle is VRFConsumerBaseV2Plus,AutomationCompatibleInterface {
    error Raffle__SendMoreToEnterRaffle();
    error Raffle__TransferFailed();
    error Raffle__NotOpen();
    enum RaffleState {
        OPEN,
        CALCULATING
    }

    event WinnerPicked(address winner);
    event RaffleEntered(address indexed player);

    uint256 immutable i_entranceFee;
    uint256 immutable i_interval;
    bytes32 immutable i_keyHash;
    uint256 immutable i_subscriptionId;
    uint16 constant REQUEST_CONFIRMATIONS = 3;
    uint32 immutable i_callbackGasLimit;
    address payable[] private s_players;
    uint256 private s_lastTimeStamp;
    uint32 constant NUM_WORDS = 2;
    address private s_recentWinner;
    RaffleState private s_raffleState;
    struct RandomWordsRequest {
        bytes32 keyHash;
        uint256 subId;
        uint16 requestConfirmations;
        uint32 callbackGasLimit;
        uint32 numWords;
        bytes extraArgs;
    }

    constructor(
        uint entranceFee,
        uint interval,
        address vrfCoordinator,
        bytes32 gasLane,
        uint256 subscriptionId,
        uint32 callbackGasLimit
    ) VRFConsumerBaseV2Plus(vrfCoordinator) {
        i_entranceFee = entranceFee;
        i_interval = interval;
        s_lastTimeStamp = block.timestamp;
        i_keyHash = gasLane;
        i_subscriptionId = subscriptionId;
        i_callbackGasLimit = callbackGasLimit;
        s_raffleState = RaffleState.OPEN;
    }

    function enterRaffle() external payable {
        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__NotOpen();
        }
        // address sender = msg.sender;
        // sender; //
        // require(msg.value >=i_entranceFee,"not sufficent funds to enter");
        //   require(msg.value >=i_entranceFee,Raffle_SendMoreToEnterRaffle());
        // most gas efficient
        if (msg.value < i_entranceFee) {
            revert Raffle__SendMoreToEnterRaffle();
        }
        s_players.push(payable(msg.sender));
        emit RaffleEntered(msg.sender);
    }

       function checkUpkeep(
        bytes calldata /* checkData */
    )
        external
        view
        override
        returns (bool upkeepNeeded, bytes memory /* performData */)
    {
        upkeepNeeded = (block.timestamp - s_lastTimeStamp) > i_interval &&s_raffleState==RaffleState.OPEN&&address(this).balance>0&&s_players.length>0;
        return (upkeepNeeded,"");
        // We don't use the checkData in this example. The checkData is defined when the Upkeep was registered.
    }

     function performUpkeep(bytes calldata /* performData */) external override {
        if ((block.timestamp - s_lastTimeStamp) > i_interval) {
            s_lastTimeStamp = block.timestamp;
            // counter = counter + 1;
            s_raffleState = RaffleState.CALCULATING;
        //picking winner with help of vrf
        VRFV2PlusClient.RandomWordsRequest memory request = VRFV2PlusClient
            .RandomWordsRequest({
                keyHash: i_keyHash,
                subId: i_subscriptionId,
                requestConfirmations: REQUEST_CONFIRMATIONS,
                callbackGasLimit: i_callbackGasLimit,
                numWords: NUM_WORDS,
                extraArgs: VRFV2PlusClient._argsToBytes(
                    // Set nativePayment to true to pay for VRF requests with Sepolia ETH instead of LINK
                    VRFV2PlusClient.ExtraArgsV1({nativePayment: false})
                )
            });
        uint256 requestId = s_vrfCoordinator.requestRandomWords(request);
        }
        // We don't use the performData in this example. The performData is generated by the Automation Node's call to your checkUpkeep function
    }

    function pickWinner() external {
        // pick winner in regular intervals
        // check if interval has passed

        //pickWinner();

        if (block.timestamp - s_lastTimeStamp < i_interval) {
            revert(" time has not passed");
        }
        s_raffleState = RaffleState.CALCULATING;
        //picking winner with help of vrf
        VRFV2PlusClient.RandomWordsRequest memory request = VRFV2PlusClient
            .RandomWordsRequest({
                keyHash: i_keyHash,
                subId: i_subscriptionId,
                requestConfirmations: REQUEST_CONFIRMATIONS,
                callbackGasLimit: i_callbackGasLimit,
                numWords: NUM_WORDS,
                extraArgs: VRFV2PlusClient._argsToBytes(
                    // Set nativePayment to true to pay for VRF requests with Sepolia ETH instead of LINK
                    VRFV2PlusClient.ExtraArgsV1({nativePayment: false})
                )
            });
        uint256 requestId = s_vrfCoordinator.requestRandomWords(request);
    }

    function getEntranceFee() external view returns (uint) {
        return i_entranceFee;
    }

    function fulfillRandomWords(
        uint256 requestId,
        uint256[] calldata randomWords
    ) internal override {
        uint indexofWinner = randomWords[0] % s_players.length;
        address payable recentWinner = payable(s_players[indexofWinner]);
        s_recentWinner = recentWinner;

        s_raffleState = RaffleState.OPEN;
        s_players = new address payable[](0);
        s_lastTimeStamp = block.timestamp;

        //   s_recentWinner.send(address(this).balance)
        (bool success, ) = s_recentWinner.call{value: address(this).balance}(
            ""
        );
        // require(success,"transfer failed");
        if (!success) {
            revert Raffle__TransferFailed();
        }

        emit WinnerPicked(recentWinner);

        //   s_raffleState=RaffleState.OPEN;
        //   s_players= new address payable[](0) ;
        //   s_lastTimeStamp=block.timestamp;
    }

    function getRaffleState() external returns(RaffleState){
        return s_raffleState;
    }
    function getRecentEntered() public returns(address){
        return s_players[s_players.length -1];
    }
}
```
HELPER CONFIG.SOL
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script} from "../lib/forge-std/src/Script.sol";
import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2Mock.sol";
abstract contract CONSTANTS {
   uint256 public constant ETH_SEPOLIA_CHAIN_ID=11155111;
   uint256 public constant LOCAL_CHAIN_ID=31337;
           uint96 baseFee = 0.25 ether; // To be understood as 0.25 LINK
        uint96 gasPriceLink = 1e9; // 1 gwei LINK

       
}

contract HelperConfig is Script,CONSTANTS{

    error InvalidChainId();
     struct NetworkConfig{
         uint entranceFee;
        uint interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint256 subscriptionId;
        uint32 callbackGasLimit;
     }
     NetworkConfig localNetworkConfig;
     mapping(uint chainId=>NetworkConfig) networkConfigs;
      constructor(){
        networkConfigs[ETH_SEPOLIA_CHAIN_ID]=getSepoliaEthConfig();
      }
      
      function getNetworkByChainId(uint256 chainId) external returns(NetworkConfig memory){
          if(networkConfigs[chainId].vrfCoordinator!=address(0)) return networkConfigs[chainId];
          else if(chainId==LOCAL_CHAIN_ID){
            return getOrCreateAnvilEthConfig();
          }
          else {
            revert InvalidChainId();
          }
      }

    function getSepoliaEthConfig() public pure returns(NetworkConfig memory ){
        return NetworkConfig({
           entranceFee:0.01 ether,
           interval:30,
           vrfCoordinator:0x9DdfaCa8183c41ad55329BdeeD9F6A8d53168B1B,
           gasLane:0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
           callbackGasLimit:500000,
           subscriptionId:0
        });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory){
        if(localNetworkConfig.vrfCoordinator!=address(0)) return localNetworkConfig;
        else {
            //  vm.startBroadcast();
        VRFCoordinatorV2Mock vrfCoordinatorV2Mock = new VRFCoordinatorV2Mock(
            baseFee,
            gasPriceLink

        );
        return localNetworkConfig= NetworkConfig({
            entranceFee: 0.01 ether,
            interval: 30, // 30 seconds
            vrfCoordinator: address(vrfCoordinatorV2Mock),
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            callbackGasLimit: 500000, // 500,000 gas
            subscriptionId: 0 // If left as 0, our scripts will create one!

        });
        }
    }
}

```
Error  while testing testDontAllowWhileCalculating()
```
 RaffleTest::testDontAllowWhileCalculating()
    ├─ [0] VM::warp(32)
    │   └─ ← [Return]
    ├─ [0] VM::roll(2)
    │   └─ ← [Return]
    ├─ [33812] Raffle::performUpkeep(0x)
    │   ├─ [258] VRFCoordinatorV2Mock::requestRandomWords(RandomWordsRequest({ keyHash: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae, subId: 0, requestConfirmations: 3, callbackGasLimit: 500000 [5e5], numWords: 2, extraArgs: 0x92fd13380000000000000000000000000000000000000000000000000000000000000000 }))
    │   │   └─ ← [Revert] EvmError: Revert
    │   └─ ← [Revert] EvmError: Revert
    └─ ← [Revert] EvmError: Revert

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 3.08ms (357.70µs CPU time)

**Comment by EngrPips** [comment]
Your `testDontAllowWhileCalculating` doesn't seem correct to me. Below is mine, which works fine.

```js
function testCantEnterWhenRaffleIsCalculating() public {
        vm.prank(player);
        raffle.enter{value: _entryFee}();
        vm.warp(block.timestamp + _interval + 1);
        vm.roll(block.number + 1);

        raffle.performUpkeep("0x0");
        console2.log(address(this));
        vm.expectRevert(Raffle.Raffle__LotteryIsCalculatingWinner.selector);
        vm.prank(player2);
        raffle.enter{value: _entryFee}();
    }
```

Also, why is your `SubId` `0` in your stack trace when you are requesting a random number? Look at the stack trace below.

```js
├─ [258] VRFCoordinatorV2Mock::requestRandomWords(RandomWordsRequest({ keyHash: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae, subId: 0, requestConfirmations: 3, callbackGasLimit: 500000 [5e5], numWords: 2, extraArgs:
```

**Reply by achyuta22** [reply]
resolved bro thankyou.

**Reply by EngrPips** [reply]
You are welcome. Keep crushing it.
